src/ray/object_manager/spilled_object.cc

```diff
@@ -0,0 +1,195 @@
+}
+
+uint64_t SpilledObject::GetDataSize() const { return data_size_; }
+
+uint64_t SpilledObject::GetMetadataSize() const { return metadata_size_; }
+
+const rpc::Address &SpilledObject::GetOwnerAddress() const { return owner_address_; }
+
+uint64_t SpilledObject::GetNumChunks() const {
+  return (data_size_ + metadata_size_ + chunk_size_ - 1) / chunk_size_;
+}
+
+absl::optional<std::string> SpilledObject::GetChunk(uint64_t chunk_index) const {
+  // The spilled file stores metadata before data. But the GetChunk needs to
+  // return data before metadata. We achieve by first read from data section,
+  // then read from metadata section.
+  auto cur_chunk_offset = chunk_index * chunk_size_;
+  auto cur_chunk_size =
+      std::min(chunk_size_, data_size_ + metadata_size_ - cur_chunk_offset);
+
+  std::string result(cur_chunk_size, '\0');
+  size_t result_offset = 0;
+
+  if (cur_chunk_offset < data_size_) {
+    // read from data section.
+    auto offset = cur_chunk_offset;
+    auto size = std::min(data_size_ - cur_chunk_offset, cur_chunk_size);
+    if (!ReadFromDataSection(offset, size, &result[result_offset])) {
+      return absl::optional<std::string>();
+    }
+    result_offset = size;
+  }
+
+  if (cur_chunk_offset + cur_chunk_size > data_size_) {
+    // read from metadata section.
+    auto offset = std::max(cur_chunk_offset, data_size_) - data_size_;
+    auto size = std::min(cur_chunk_offset + cur_chunk_size - data_size_, cur_chunk_size);
+    if (!ReadFromMetadataSection(offset, size, &result[result_offset])) {
+      return absl::optional<std::string>();
+    }
+  }
+  return absl::optional<std::string>(std::move(result));
+}
+
+SpilledObject::SpilledObject(std::string file_path, uint64_t object_size,
+                             uint64_t data_offset, uint64_t data_size,
+                             uint64_t metadata_offset, uint64_t metadata_size,
+                             rpc::Address owner_address, uint64_t chunk_size)
+    : file_path_(std::move(file_path)),
+      object_size_(object_size),
+      data_offset_(data_offset),
+      data_size_(data_size),
+      metadata_offset_(metadata_offset),
+      metadata_size_(metadata_size),
+      owner_address_(std::move(owner_address)),
+      chunk_size_(chunk_size) {}
+
+/* static */ bool SpilledObject::ParseObjectURL(const std::string &object_url,
+                                                std::string &file_path,
+                                                uint64_t &object_offset,
+                                                uint64_t &object_size) {
+  static const std::regex object_url_pattern("^(.*)\\?offset=(\\d+)&size=(\\d+)$");
+  std::smatch match_groups;
+  if (!std::regex_match(object_url, match_groups, object_url_pattern) ||
+      match_groups.size() != 4) {
+    return false;
+  }
+  file_path = match_groups[1].str();
+  try {
+    object_offset = std::stoi(match_groups[2].str());
+    object_size = std::stoi(match_groups[3].str());
```

Comment:
yup. i think the question is around when failure happens how we want the system proceed. should it crash or detect the error and proceed? 
my general strategy here is to not crashing but report error, and let caller proceed.

Reply:
But if the file format is changed, we just need to fix it, and it seems uncommon this will raise any exception in other cases. Isn't this eventually making it harder to find bugs if formats are changed? I think we generally use strong check in this sort of things.